### PR TITLE
DOCSUP-13742: Support PARTITION BY for the s3 table function

### PR DESCRIPTION
**Pull request by olgarev**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Documentation for #23051

### CODE REVIEW

File: docs/ru/sql-reference/table-functions/s3.md [hunk 1]
```diff
@@ -133,6 +133,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
 SELECT name, value FROM existing_table;
 ```
 
+## Партиционирование при записи данных {#partitioned-write}
+
+При добавлении данных в таблицу S3 вы можете сохранить таблицу в нескольких файлах, указав выражение `PARTITION BY`. В этом случае для каждого значения выражения партиционирования создается отдельный файл. Сохранение таблицы в нескольких отдельных файлах повышает эффективность операций чтения.
```
Comment (adevyatova) [inline docs/ru/sql-reference/table-functions/s3.md:138]:
```suggestion
Если при добавлении данных в таблицу S3 вы укажете выражение `PARTITION BY`, то для каждого значения выражения партиционирования будет создан отдельный файл. Это повышает эффективность операций чтения.
```

File: docs/en/sql-reference/table-functions/s3.md [hunk 2]
```diff
@@ -125,6 +125,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
 SELECT name, value FROM existing_table;
 ```
 
+## Partitioned Write {#partitioned-write}
+
+When inserting data into S3 table, you can split it into several files by specifying `PARTITION BY` expression. In this case a separate file is created for each partition value. Splitting the data helps to improve reading operations efficiency.
```
Comment (adevyatova) [inline docs/en/sql-reference/table-functions/s3.md:130]:
```suggestion
If you will specify `PARTITION BY` expression when inserting data into `S3` table, a separate file is created for each partition value. Splitting the data helps to improve reading operations efficiency.
```

File: docs/en/sql-reference/table-functions/s3.md [hunk 3]
```diff
@@ -125,6 +125,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
 SELECT name, value FROM existing_table;
 ```
 
+## Partitioned Write {#partitioned-write}
+
+If you will specify `PARTITION BY` expression when inserting data into `S3` table, a separate file is created for each partition value. Splitting the data helps to improve reading operations efficiency.
```
Comment (olgarev) [inline docs/en/sql-reference/table-functions/s3.md:130]:
```suggestion
If you specify `PARTITION BY` expression when inserting data into `S3` table, a separate file is created for each partition value. Splitting the data helps to improve reading operations efficiency.
```

File: docs/ru/sql-reference/table-functions/s3.md [hunk 4]
```diff
@@ -133,6 +133,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
 SELECT name, value FROM existing_table;
 ```
 
+## Партиционирование при записи данных {#partitioned-write}
+
+Если при добавлении данных в таблицу S3 вы укажете выражение `PARTITION BY`, то для каждого значения выражения партиционирования будет создан отдельный файл. Это повышает эффективность операций чтения.
```
Comment (olgarev) [inline docs/ru/sql-reference/table-functions/s3.md:138]:
```suggestion
Если при добавлении данных в таблицу S3 указать выражение `PARTITION BY`, то для каждого значения выражения партиционирования создается отдельный файл. Это повышает эффективность операций чтения.
```

File: docs/en/sql-reference/table-functions/s3.md [hunk 5]
```diff
@@ -125,6 +125,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
 SELECT name, value FROM existing_table;
 ```
 
+## Partitioned Write {#partitioned-write}
+
+If you specify `PARTITION BY` expression when inserting data into `S3` table, a separate file is created for each partition value. Splitting the data helps to improve reading operations efficiency.
```
Comment (olgarev) [inline docs/en/sql-reference/table-functions/s3.md:130]:
```suggestion
If you specify `PARTITION BY` expression when inserting data into `S3` table, a separate file is created for each partition value. Splitting the data into separate files helps to improve reading operations efficiency.
```

File: docs/en/sql-reference/table-functions/s3.md [hunk 6]
```diff
@@ -125,6 +125,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
...
+If you specify `PARTITION BY` expression when inserting data into `S3` table, a separate file is created for each partition value. Splitting the data into separate files helps to improve reading operations efficiency.
+
+**Examples**
+
+1. Using partition ID in a file name creates separate files:
+
+```sql
+INSERT INTO TABLE FUNCTION
+    s3('http://bucket.amazonaws.com/my_bucket/file_{_partition_id}.csv', 'CSV', 'a UInt32, b UInt32, c UInt32')
+    PARTITION BY a VALUES (1, 2, 3), (1, 4, 5), (10, 11, 12), (10, 13, 14), (20, 21, 22), (20, 23, 24)
```
Comment (olgarev) [inline docs/en/sql-reference/table-functions/s3.md:139]:
```suggestion
INSERT INTO TABLE FUNCTION
    s3('http://bucket.amazonaws.com/my_bucket/file_{_partition_id}.csv', 'CSV', 'a String, b UInt32, c UInt32')
    PARTITION BY a VALUES ('x', 2, 3), ('x', 4, 5), ('y', 11, 12), ('y', 13, 14), ('z', 21, 22), ('z', 23, 24)
```

File: docs/en/sql-reference/table-functions/s3.md [hunk 7]
```diff
@@ -125,6 +125,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
...
+```
+As a result, the data is written into three files: `file_1.csv`, `file_10.csv`, and `file_20.csv`.
```
Comment (olgarev) [inline docs/en/sql-reference/table-functions/s3.md:141]:
```suggestion
As a result, the data is written into three files: `file_x.csv`, `file_y.csv`, and `file_z.csv`.
```

File: docs/ru/sql-reference/table-functions/s3.md [hunk 8]
```diff
@@ -133,6 +133,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
...
+Если при добавлении данных в таблицу S3 указать выражение `PARTITION BY`, то для каждого значения выражения партиционирования создается отдельный файл. Это повышает эффективность операций чтения.
+
+**Примеры**
+
+1. При использовании ID партиции в названии файла создаются отдельные файлы:
+
+```sql
+INSERT INTO TABLE FUNCTION
+    s3('http://bucket.amazonaws.com/my_bucket/file_{_partition_id}.csv', 'CSV', 'a UInt32, b UInt32, c UInt32')
+    PARTITION BY a VALUES (1, 2, 3), (1, 4, 5), (10, 11, 12), (10, 13, 14), (20, 21, 22), (20, 23, 24)
```
Comment (olgarev) [inline docs/ru/sql-reference/table-functions/s3.md:147]:
```suggestion
INSERT INTO TABLE FUNCTION
    s3('http://bucket.amazonaws.com/my_bucket/file_{_partition_id}.csv', 'CSV', 'a UInt32, b UInt32, c UInt32')
    PARTITION BY a VALUES ('x', 2, 3), ('x', 4, 5), ('y', 11, 12), ('y', 13, 14), ('z', 21, 22), ('z', 23, 24)
```

File: docs/en/sql-reference/table-functions/s3.md [hunk 9]
```diff
@@ -125,6 +125,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
...
+
+2. Using partition ID in a bucket name creates files in different buckets:
+
+```sql
+INSERT INTO TABLE FUNCTION
+    s3('http://bucket.amazonaws.com/my_bucket_{_partition_id}/file.csv', 'CSV', 'a UInt32, b UInt32, c UInt32')
+    PARTITION BY a VALUES (1, 2, 3), (1, 4, 5), (10, 11, 12), (10, 13, 14), (20, 21, 22), (20, 23, 24)
```
Comment (olgarev) [inline docs/en/sql-reference/table-functions/s3.md:148]:
```suggestion
    PARTITION BY a VALUES (1, 2, 3), (1, 4, 5), (10, 11, 12), (10, 13, 14), (20, 21, 22), (20, 23, 24);
```

File: docs/ru/sql-reference/table-functions/s3.md [hunk 10]
```diff
@@ -133,6 +133,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
...
+```
+В результате данные будут записаны в три файла: `file_1.csv`, `file_10.csv` и `file_20.csv`.
```
Comment (olgarev) [inline docs/ru/sql-reference/table-functions/s3.md:149]:
```suggestion
В результате данные будут записаны в три файла: `file_x.csv`, `file_y.csv` и `file_z.csv`.
```

File: docs/ru/sql-reference/table-functions/s3.md [hunk 11]
```diff
@@ -133,6 +133,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
...
+
+2. При использовании ID партиции в названии бакета создаются файлы в разных бакетах:
+
+```sql
+INSERT INTO TABLE FUNCTION
+    s3('http://bucket.amazonaws.com/my_bucket_{_partition_id}/file.csv', 'CSV', 'a UInt32, b UInt32, c UInt32')
+    PARTITION BY a VALUES (1, 2, 3), (1, 4, 5), (10, 11, 12), (10, 13, 14), (20, 21, 22), (20, 23, 24)
```
Comment (olgarev) [inline docs/ru/sql-reference/table-functions/s3.md:156]:
```suggestion
    PARTITION BY a VALUES (1, 2, 3), (1, 4, 5), (10, 11, 12), (10, 13, 14), (20, 21, 22), (20, 23, 24);
```

**Comment by olgarev** [comment]
@vdimir Hi, please could you review this PR?

**Comment by alex-zaitsev** [comment]
@excitoon , please look

File: docs/en/sql-reference/table-functions/s3.md [hunk 12]
```diff
@@ -125,6 +125,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
 SELECT name, value FROM existing_table;
 ```
 
+## Partitioned Write {#partitioned-write}
+
+If you specify `PARTITION BY` expression when inserting data into `S3` table, a separate file is created for each partition value. Splitting the data into separate files helps to improve reading operations efficiency.
+
+**Examples**
+
+1. Using partition ID in a file name creates separate files:
```
Comment (vdimir) [inline docs/en/sql-reference/table-functions/s3.md:134]:
```suggestion
1. Using partition ID in a key creates separate objects:
```

File: docs/en/sql-reference/table-functions/s3.md [hunk 13]
```diff
@@ -125,6 +125,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
...
+
+```sql
+INSERT INTO TABLE FUNCTION
+    s3('http://bucket.amazonaws.com/my_bucket/file_{_partition_id}.csv', 'CSV', 'a String, b UInt32, c UInt32')
+    PARTITION BY a VALUES ('x', 2, 3), ('x', 4, 5), ('y', 11, 12), ('y', 13, 14), ('z', 21, 22), ('z', 23, 24);
+```
+As a result, the data is written into three files: `file_x.csv`, `file_y.csv`, and `file_z.csv`.
```
Comment (vdimir) [inline docs/en/sql-reference/table-functions/s3.md:141]:
```suggestion
As a result, the data is written into three objects: `file_x.csv`, `file_y.csv`, and `file_z.csv`.
```

File: docs/en/sql-reference/table-functions/s3.md [hunk 14]
```diff
@@ -125,6 +125,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
...
+
+2. Using partition ID in a bucket name creates files in different buckets:
```
Comment (vdimir) [inline docs/en/sql-reference/table-functions/s3.md:143]:
```suggestion
2. Using partition ID in a bucket name creates objects in different buckets:
```

File: docs/en/sql-reference/table-functions/s3.md [hunk 15]
```diff
@@ -125,6 +125,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
...
+
+```sql
+INSERT INTO TABLE FUNCTION
+    s3('http://bucket.amazonaws.com/my_bucket_{_partition_id}/file.csv', 'CSV', 'a UInt32, b UInt32, c UInt32')
+    PARTITION BY a VALUES (1, 2, 3), (1, 4, 5), (10, 11, 12), (10, 13, 14), (20, 21, 22), (20, 23, 24);
+```
+As a result, the data is written into three files in different buckets: `my_bucket_1/file.csv`, `my_bucket_10/file.csv`, and `my_bucket_20/file.csv`.
```
Comment (vdimir) [inline docs/en/sql-reference/table-functions/s3.md:150]:
```suggestion
As a result, the data is written into three objects in different buckets: `my_bucket_1/file.csv`, `my_bucket_10/file.csv`, and `my_bucket_20/file.csv`.
```

**Comment by excitoon** [comment]
LGTM

File: docs/en/sql-reference/table-functions/s3.md [hunk 16]
```diff
@@ -125,6 +125,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
 SELECT name, value FROM existing_table;
 ```
 
+## Partitioned Write {#partitioned-write}
+
+If you specify `PARTITION BY` expression when inserting data into `S3` table, a separate file is created for each partition value. Splitting the data into separate files helps to improve reading operations efficiency.
```
Comment (vdimir) [inline docs/en/sql-reference/table-functions/s3.md:130]:
```suggestion
If you specify `PARTITION BY` expression when inserting data into `S3` table, a separate objects is created for each partition value. Splitting the data into separate objects helps to improve reading operations efficiency.
```

File: docs/ru/sql-reference/table-functions/s3.md [hunk 17]
```diff
@@ -133,6 +133,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
 SELECT name, value FROM existing_table;
 ```
 
+## Партиционирование при записи данных {#partitioned-write}
+
+Если при добавлении данных в таблицу S3 указать выражение `PARTITION BY`, то для каждого значения выражения партиционирования создается отдельный файл. Это повышает эффективность операций чтения.
```
Comment (vdimir) [inline docs/ru/sql-reference/table-functions/s3.md:138]:
```suggestion
Если при добавлении данных в таблицу S3 указать выражение `PARTITION BY`, то для каждого значения ключа партиционирования создается отдельный объект. Это повышает эффективность операций чтения.
```

File: docs/ru/sql-reference/table-functions/s3.md [hunk 18]
```diff
@@ -133,6 +133,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
...
+
+**Примеры**
+
+1. При использовании ID партиции в названии файла создаются отдельные файлы:
```
Comment (vdimir) [inline docs/ru/sql-reference/table-functions/s3.md:142]:
```suggestion
1. При использовании ID партиции в имени ключа создаются отдельные объекты:
```

File: docs/ru/sql-reference/table-functions/s3.md [hunk 19]
```diff
@@ -133,6 +133,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
...
+
+```sql
+INSERT INTO TABLE FUNCTION
+    s3('http://bucket.amazonaws.com/my_bucket/file_{_partition_id}.csv', 'CSV', 'a UInt32, b UInt32, c UInt32')
+    PARTITION BY a VALUES ('x', 2, 3), ('x', 4, 5), ('y', 11, 12), ('y', 13, 14), ('z', 21, 22), ('z', 23, 24);
+```
+В результате данные будут записаны в три файла: `file_x.csv`, `file_y.csv` и `file_z.csv`.
```
Comment (vdimir) [inline docs/ru/sql-reference/table-functions/s3.md:149]:
```suggestion
В результате данные будут записаны в три объекта: `file_x.csv`, `file_y.csv` и `file_z.csv`.
```

File: docs/ru/sql-reference/table-functions/s3.md [hunk 20]
```diff
@@ -133,6 +133,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
...
+
+2. При использовании ID партиции в названии бакета создаются файлы в разных бакетах:
```
Comment (vdimir) [inline docs/ru/sql-reference/table-functions/s3.md:151]:
```suggestion
2. При использовании ID партиции в названии бакета создаются объекты в разных бакетах:
```

File: docs/ru/sql-reference/table-functions/s3.md [hunk 21]
```diff
@@ -133,6 +133,30 @@ INSERT INTO FUNCTION s3('https://storage.yandexcloud.net/my-test-bucket-768/test
...
+
+```sql
+INSERT INTO TABLE FUNCTION
+    s3('http://bucket.amazonaws.com/my_bucket_{_partition_id}/file.csv', 'CSV', 'a UInt32, b UInt32, c UInt32')
+    PARTITION BY a VALUES (1, 2, 3), (1, 4, 5), (10, 11, 12), (10, 13, 14), (20, 21, 22), (20, 23, 24);
+```
+В результате будут созданы три файла в разных бакетах: `my_bucket_1/file.csv`, `my_bucket_10/file.csv` и `my_bucket_20/file.csv`.
```
Comment (vdimir) [inline docs/ru/sql-reference/table-functions/s3.md:158]:
```suggestion
В результате будут созданы три объекта в разных бакетах: `my_bucket_1/file.csv`, `my_bucket_10/file.csv` и `my_bucket_20/file.csv`.
```

**Review by vdimir — APPROVED** [review]
I've just changed `file` to `object`, seems it's more common in a context of S3 (but actually both are used)

https://aws.amazon.com/s3/faqs


_UPD:_ Actually `file` already is used in our documentation for s3-related things, so let's it's better to keep it consistent, let's leave it

File: docs/ru/sql-reference/table-functions/s3.md (hunk 17 above)
Comment (vdimir) [inline docs/ru/sql-reference/table-functions/s3.md:138]:
```suggestion
Если при добавлении данных в таблицу S3 указать выражение `PARTITION BY`, то для каждого значения ключа партиционирования создается отдельный файл. Это повышает эффективность операций чтения.
```

File: docs/ru/sql-reference/table-functions/s3.md (hunk 18 above)
Comment (vdimir) [inline docs/ru/sql-reference/table-functions/s3.md:142]:
```suggestion
1. При использовании ID партиции в имени ключа создаются отдельные файлы:
```

File: docs/en/sql-reference/table-functions/s3.md (hunk 12 above)
Comment (vdimir) [inline docs/en/sql-reference/table-functions/s3.md:134]:
```suggestion
1. Using partition ID in a key creates separate files:
```

**Comment by vdimir** [comment]
I'm going to merge this, @excitoon if you will have any notes, then will make it in other PR.

**Pull request merged**